File: cpp/2D_DP/62.unique-paths.cpp
#include "../std.hpp"

class Solution {
public:
    int uniquePaths(int m, int n)
    {
        vector<vector<int>> grid(m, vector<int>(n, 0));
        for (auto &&r : grid) {
            r[0] = 1;
        }
        for (auto &&c : grid[0]) {
            c = 1;
        }
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                grid[i][j] = grid[i - 1][j] + grid[i][j - 1];
            }
        }
        return grid[m - 1][n - 1];
    }
};

File: js/2D_DP/62.unique-paths.ts
function uniquePaths(m: number, n: number): number {
  let grid = Array.from(Array(m), () => new Array(n))
  grid[0].fill(1)
  for (let row of grid) {
    row[0] = 1
  }
  for (let i = 1; i < m; ++i) {
    for (let j = 1; j < n; ++j) {
      grid[i][j] = grid[i - 1][j] + grid[i][j - 1]
    }
  }
  return grid.at(-1).at(-1)
}
console.log(uniquePaths(3, 2))
